### PR TITLE
Normalization

### PR DESCRIPTION
**Pull request by afrittoli**

### CODE REVIEW

File: ciml/trainer.py [hunk 1]
```diff
@@ -245,6 +247,7 @@ def prepare_dataset(dataset, normalized_length, num_dstat_features, data_type,
 
         # Normalize data
         example = fixed_lenght_example(result, normalized_length)
+
```
Comment (mtreinish) [inline ciml/trainer.py:266]: Unrelated whitespace, tsk tsk

File: ciml/trainer.py [hunk 2]
```diff
@@ -372,9 +382,10 @@ def build_dataset(dataset, build_name, limit, sample_interval, features_regex,
 
     # Load available run ids for the build name
     runs = gather_results.load_run_uuids('.raw', name=build_name)
-    # Apply the limit
-    if limit > 0:
-        runs = np.array(runs[:limit])
+    # Apply the slice
+    slice_fn = lambda x: int(x.strip()) if x.strip() else None
+    slice_object = slice(*map(slice_fn, slicer.split(":")))
```
Comment (mtreinish) [inline ciml/trainer.py:403]: Do we have a docstring or something for how this works? I've been sitting here staring at it and am having a hard time visualizing in my head how this works, or how you would use the slicer arg with this.

File: ciml/trainer.py [hunk 3]
```diff
@@ -156,7 +157,7 @@ def normalize_dataset(examples, labels, params=None):
         else:
             mean_fd = np.mean(feature_data)
             max_min_fd = np.max(feature_data) - np.min(feature_data)
-            # In case of just one example, or a flat feature
+            # In case of just one example, or
```
Comment (mtreinish) [inline ciml/trainer.py]: or what? :)

File: ciml/trainer.py [hunk 4]
```diff
@@ -555,41 +596,49 @@ def local_trainer(dataset, experiment, gpu, debug):
     training_data = gather_results.load_dataset(dataset, 'training')
     test_data = gather_results.load_dataset(dataset, 'test')
     print("Training data shape: (%d, %d)" % training_data['examples'].shape)
-    print("Evaluation data shape: (%d, %d)" % test_data['examples'].shape)
+
+    if class_label == 'node_provider':
+        label_vocabulary = set(['rax-iad', 'ovh-bhs1', 'packethost-us-west-1',
```
Comment (mtreinish) [inline ciml/trainer.py:601]: This limits portability a bit. But we're not really at a place where that really matters.

File: ciml/trainer.py [hunk 5]
```diff
@@ -555,41 +596,49 @@ def local_trainer(dataset, experiment, gpu, debug):
...
+            # Saving and Logging loss
+            print('Training eval data for %s: %r' % (eval_dataset_name,
+                                                     eval_loss))
+            eval_name = "eval_" + eval_dataset_name
+            gather_results.save_data_json(dataset, eval_loss, eval_name,
+                                          sub_folder=experiment)
+
+    # Now do the training and evalutation
+    if gpu:
+        with tf.device('/device:GPU:0'):
```
Comment (mtreinish) [inline ciml/trainer.py]: FWIW, when I was doing my local testing I don't think this really made a difference. The first gpu still wasn't being leveraged on my system. I realize this is just migrated from above, but just wanted to point that out as I went through everything. Although, maybe with the other changes tensorflow will decide that it can actually use the gpu now.

**Review by mtreinish — APPROVED** [review]
LGTM, just some inline comments. Nothing worth blocking over though.

File: ciml/trainer.py (hunk 1 above)
Comment (afrittoli) [reply ciml/trainer.py:266]: :P

File: ciml/trainer.py (hunk 2 above)
Comment (afrittoli) [reply ciml/trainer.py:403]: Heh, ok I thought the help text would be enough, I will add some docstring.
Basically you can setup specify any slice like you would do in an array e.g.

10:-2:4 -> start at element 10 until 2 elements before the end with pace 4

File: ciml/trainer.py (hunk 4 above)
Comment (afrittoli) [reply ciml/trainer.py:601]: Well spotted, this was a quick fix to get things moving. I will improve on that.

File: ciml/trainer.py (hunk 2 above)
Comment (mtreinish) [reply ciml/trainer.py:403]: Ah ok, that's what I thought based on the variable names and the default value for slicer arg. It just wasn't clear to me from the code (I guess I could have just spun up the interpreter to test it) by looking at it.
